src/java/org/apache/cassandra/hints/HintMessage.java

```diff
@@ -74,15 +75,76 @@
         this.unknownTableID = unknownTableID;
     }
 
+    // Fake serializer for dtests
+    @VisibleForTesting
+    public static class DTestSerializer implements IVersionedAsymmetricSerializer<SerializableHintMessage, HintMessage>
+    {
+
+        public void serialize(SerializableHintMessage obj, DataOutputPlus out, int version) throws IOException
+        {
+            if (obj instanceof HintMessage)
+            {
+                HintMessage message = (HintMessage) obj;
+                if (message.hint == null)
+                {
+                    UUIDSerializer.serializer.serialize(message.hostId, out, version);
+                    out.writeUnsignedVInt(0);
+                    message.unknownTableID.serialize(out);
+                    return;
+                }
+            }
+
+            serializer.serialize(obj, out, version);
+        }
+
+        public HintMessage deserialize(DataInputPlus in, int version) throws IOException
```

Comment:
Why not delegate to the real serializer for `deserialize` like you do for `serialized`/`serializedSize` as it should already tolerate missing tables?

Reply:
the problem is not in tolerating missing tables, but in the fact that we're re-serializing `HintMessage` where `hint` is `null`

Reply:
understood, I meant why not this
```
        public HintMessage deserialize(DataInputPlus in, int version) throws IOException
        {
            return serializer.deserialize(in, version);
        }
```

but hadn't realized you were relying on checking hintSize==0 rather than the existing UnknownTableException logic.